src/Raven.Server/Documents/Queries/Sorting/AlphaNumeric/AlphaNumericFieldComparator.cs

```diff
@@ -99,99 +100,110 @@ internal sealed class AlphanumComparer : IComparer<UnmanagedStringArray.Unmanage
 
-                private static int ReadOneChar(Span<byte> str, int offset, ref char ch)
+                private static (int BytesUsed, int CharUsed) ReadCharacter(ReadOnlySpan<byte> str, int offset, Span<char> charactersBuffer)
                 {
-                    fixed (byte* buffer = str)
-                    fixed (char* c = &ch)
+                    var decoder = Decoder ??= Encoding.UTF8.GetDecoder();
+                    
+                    //Numbers and ASCII are always 1 so we will pay the price only in case of UTF-8 characters.
+                    //http://www.unicode.org/versions/Unicode9.0.0/ch03.pdf#page=54
+                    var (byteLengthOfCharacter, charNeededToEncodeCharacters) = str[offset] switch
                     {
-                        var decoder = Decoder ??= Encoding.UTF8.GetDecoder();
-                        decoder.Convert(buffer + offset, str.Length - offset, c, 1, flush: true, out var bytesUsed,
-                            out var charUsed, out _);
-
-                        if (charUsed != 1)
-                            throw new InvalidOperationException($"Read unexpected number of chars {charUsed} from string: '{Encoding.UTF8.GetString(str)}' at offset: {offset}");
-
-                        return bytesUsed;
-                    }
+                        <= 0b0111_1111 => (1,1), /* 1 byte sequence: 0b0xxxxxxxx */
+                        <= 0b1101_1111 => (2, Encoding.UTF8.GetCharCount(str.Slice(offset, 2))), /* 2 byte sequence: 0b110xxxxxx */
+                        <= 0b1110_1111 => (3, Encoding.UTF8.GetCharCount(str.Slice(offset, 3))), /* 0b1110xxxx: 3 bytes sequence */
+                        <= 0b1111_0111 => (4, Encoding.UTF8.GetCharCount(str.Slice(offset, 4))), /* 0b11110xxx: 4 bytes sequence */
+                        _ => throw new InvalidDataException($"Characters should be between 1 and 4 bytes long and cannot match the specified sequence. This is invalid code.")
+                    };
+                    
+                    Debug.Assert(charactersBuffer.Length >= charNeededToEncodeCharacters, $"Character requires more than {charactersBuffer.Length} space to be decoded.");
+                    
+                    //In case of surrogate we could've to use two characters to convert it.
+                    decoder.Convert(str.Slice(offset, byteLengthOfCharacter), charactersBuffer, flush: true, out int bytesUsed,out int charUsed, out _);
+                    
+                    return (bytesUsed, charUsed);
                 }
 
-                public int CompareWithAnotherState(AlphanumericStringComparisonState other)
+                public int CompareWithAnotherState(ref AlphanumericStringComparisonState other)
                 {
-                    var string1State = this;
-                    var string2State = other;
+                    ref var string1State = ref this;
+                    ref var string2State = ref other;
```

Comment:
I think there is no need to copy the whole structs for each state comparison.

Reply:
+1